### PR TITLE
Tc scheduler rewrite

### PR DESCRIPTION
**Pull request by barath**
This is a large pull request rewriting the traffic class and scheduler code, and related functionality.

### CODE REVIEW

**Comment by melvinw** [comment]
pls squash-and-merge :)

File: bessctl/commands.py [hunk 1]
```diff
@@ -822,14 +821,14 @@ def _show_tc_list(cli, tcs):
...
-                           '%s\n' %
-                           (getattr(tc, 'class').name,
-                            tc.parent if tc.parent else 'none',
-                            getattr(tc, 'class').priority,
-                            tc.tasks,
-                            _limit_to_str(getattr(tc, 'class').limit)))
+            cli.fout.write('    %-16s  ' \
+                           'parent %-10s  priority %-3d  tasks %-3d ' \
+                           '%s\n' % \
+                    (getattr(tc, 'class').name,
```
Comment (changlan) [inline bessctl/commands.py]: Nit: the following lines are under-indented

File: core/Makefile [hunk 2]
```diff
@@ -55,6 +55,9 @@ CXXFLAGS += -std=gnu++11 -g3 -ggdb3 -Ofast -march=native \
             -Wall -Wextra -Wcast-align
 
 PERMISSIVE = -Wno-unused-parameter
+CXXFLAGS += -std=gnu++11 -g3 -ggdb3 -Og -march=native \
```
Comment (changlan) [inline core/Makefile]: Redundant

File: core/Makefile [hunk 3]
```diff
@@ -67,14 +70,14 @@ else
...
-           -static-libstdc++
+					 -static-libstdc++ 
 
 LIBS += -Wl,-non_shared \
 	-Wl,--whole-archive -l$(DPDK_LIB) -Wl,--no-whole-archive \
 	-lglog -lgflags -lprotobuf -lgrpc++ -lgrpc \
 	-lssl -lcrypto -lunwind -llzma -lpcap -lz \
 	-Wl,-call_shared \
-	-ldl
+	-ldl
```
Comment (changlan) [inline core/Makefile]: Trailing space

File: core/Makefile [hunk 4]
```diff
@@ -67,14 +70,14 @@ else
 endif
 
 LDFLAGS += -rdynamic -L$(DPDK_LIB_DIR) -Wl,-rpath=$(DPDK_LIB_DIR) -pthread \
-           -static-libstdc++
+					 -static-libstdc++
```
Comment (changlan) [inline core/Makefile]: Indentation and trailing space

File: core/Makefile [hunk 5]
```diff
@@ -88,8 +91,8 @@ ifdef COVERAGE
 endif
 
 ifndef DEBUG
-    CXXFLAGS += -Ofast -DNDEBUG
-    LDFLAGS += -ffat-lto-objects -finline-limit=1000
+    CXXFLAGS += -Ofast -DNDEBUG -flto -ffat-lto-objects -finline-limit=1000
```
Comment (changlan) [inline core/Makefile]: -flto won't be necessary in CXXFLAGS?

File: core/Makefile [hunk 6]
```diff
@@ -88,8 +91,8 @@ ifdef COVERAGE
...
+    LDFLAGS += -flto -ffat-lto-objects -finline-limit=1000
```
Comment (changlan) [inline core/Makefile]: Only add -flto when the compiler is not Clang (as we did in the following lines), so we can remove -flto in this line.

File: core/bessctl.cc [hunk 7]
```diff
@@ -427,51 +424,40 @@ class BESSControlImpl final : public BESSControl::Service {
...
-      status->mutable_class_()->mutable_max_burst()->set_bits(
-          c->settings.max_burst[3]);
+      status->mutable_class_()->set_wid(wid_filter);
+
+      if (c->policy() == bess::POLICY_RATE_LIMIT) {
+        bess::RateLimitTrafficClass *rl = reinterpret_cast<bess::RateLimitTrafficClass*>(c);
+        std::string resource = bess::ResourceName.at(rl->resource());
+        int64_t limit = rl->limit();
+        int64_t max_burst = rl->max_burst();
+        status->mutable_class_()->mutable_limit()->insert({resource, limit});
```
Comment (changlan) [inline core/bessctl.cc]: Much simpler! 👍

File: core/traffic_class.h [hunk 8]
```diff
@@ -0,0 +1,645 @@
...
+
+#define SCHED_DEBUG 0
+
+// A large default priority.
+#define DEFAULT_PRIORITY 0xFFFFFFFFu
+
+#define MAX_LIMIT_POW 36
+#define USAGE_AMPLIFIER_POW 32
+
+/* share is defined relatively, so 1024 should be large enough */
```
Comment (changlan) [inline core/traffic_class.h]: Nit: Perhaps use // so that the style is consistent?

File: core/traffic_class.h [hunk 9]
```diff
@@ -0,0 +1,645 @@
...
+#define MAX_SHARE (1 << 10)
+#define STRIDE1 (1 << 20)
+
+/* this doesn't mean anything, other than avoiding int64 overflow */
```
Comment (changlan) [inline core/traffic_class.h]: Same here

File: core/traffic_class.h [hunk 10]
```diff
@@ -1,5 +1,5 @@
-#ifndef BESS_TC_H_
-#define BESS_TC_H_
+#ifndef BESS_TRAFFIC_CLASS_H_
```
Comment (sangjinhan) [inline core/traffic_class.h]: nit: existing header guards do not include "_" of the filename.

**Comment by changlan** [comment]
Any idea why the gcc 4.8 build fails?

File: core/task.h [hunk 11]
```diff
@@ -4,45 +4,37 @@
...
-
-  Module *m;
-  void *arg;
-
-  struct cdlist_item tc;
-  struct cdlist_item all_tasks;
-};
+namespace bess {
+class LeafTrafficClass;
+}  // namespace bess
```
Comment (barath) [inline core/task.h]: Basically task hasn't fully been ported.

**Comment by justinemarie** [comment]
Melvin wonderfully walked through the updates with me because he's great. I have two questions --

* Can we have a policy that, when we use auto and it's not completely obvious what the return value/type is, we put a quick comment explaining what it is? Auto makes the code 80% more concise and 90% less understandable.

* Is it really the case that modules have more than one task associated with them? Melvin and I were trying to think of examples when that is true but we couldn't think of any that weren't silly. It seems like each module should have 0 or 1 tasks associated with them -- they're either directly invoked by the scheduler, or indirectly invoked as packets are pushed down the pipeline to them. But why would they need 2+ tasks?

File: core/module.cc [hunk 12]
```diff
@@ -213,24 +213,19 @@ void Module::ProcessBatch(bess::PacketBatch *) {
 }
 
 task_id_t Module::RegisterTask(void *arg) {
-  task_id_t id;
-  struct task *t;
+  Worker *w = get_next_active_worker();
+  auto *c = w->s()->default_leaf_class();
```
Comment (justinemarie) [inline core/module.cc]: Example of the auto confusion -- is this the worker class type, or is this an actual worker instance?

**Comment by melvinw** [comment]
This g++-4.8 error is really strange. The log is reporting a segfault in `traffic_class_test`. I can't reproduce locally.

**Comment by melvinw** [comment]
That is, after installing g++-4.8, rebuilding protobuf, etc...

**Comment by melvinw** [comment]
just a reminder to squash on merge. this pr is super messy.

**Comment by changlan** [comment]
image

Merged, thanks!

**Pull request merged**